File: src/mod10_design_pattern/lesson12_strategy/java/Student.java
package mod10_design_pattern.lesson12_strategy.java;

public class Student {
    private String name;
    private StudyStrategy studyStrategy;

    public Student(String name, StudyStrategy studyStrategy) {
        this.name = name;
        this.studyStrategy = studyStrategy;
    }

    public void setStudyStrategy(StudyStrategy studyStrategy) {
        this.studyStrategy = studyStrategy;
    }

    public void learnSpell(String spell){
        System.out.println(name + " is learning:");
        studyStrategy.study(spell);
    }
}

File: src/mod10_design_pattern/lesson12_strategy/kotlin/Student.kt
package mod10_design_pattern.lesson12_strategy.kotlin

import mod10_design_pattern.lesson7_adapter.kotlin.Spell

class Student(
    private val name:String,
    private var strategy: StudyStrategy
) {

    fun setStudyStrategy(studyStrategy: StudyStrategy){
        this.strategy=studyStrategy
    }

    fun learnSpell(spell: String){
        println("$name is learning:")
        strategy.study(spell)
    }
}